### PR TITLE
Feat Booking Limits

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?

UI

CleanShot 2022-10-04 at 12 56 07

Backend Stuff
https://www.loom.com/share/d6c01ec823554d2094b377dda87449e7

Fixes #2684 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?
* run tests
* add limits 
* check bookings are blocked

### CODE REVIEW

File: packages/lib/checkBookingLimits.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  const limitCalculations = Object.entries(bookingLimits).map(
+    async ([key, limitingNumber]) => await checkLimit({ key, limitingNumber, eventStartDate, eventId })
+  );
+  await Promise.all(limitCalculations).catch((error) => {
+    throw new HttpError({ message: error.message, statusCode: 401 });
+  });
+  return true;
+}
+
+export async function checkLimit({
```
Comment (sean-brydon) [inline packages/lib/checkBookingLimits.ts:17]: Extracted so we can test with different prisma mocks

**Comment by PeerRich** [comment]
can't wait for this

File: packages/core/getUserAvailability.ts [hunk 2]
```diff
@@ -78,6 +79,23 @@ export const getCurrentSeats = (eventTypeId: number, dateFrom: Dayjs, dateTo: Da
 
 export type CurrentSeats = Awaited<ReturnType<typeof getCurrentSeats>>;
 
+export async function getBookingLimits({
```
Comment (sean-brydon) [inline packages/core/getUserAvailability.ts:82]: WIP

File: apps/web/components/v2/eventtype/EventLimitsTab.tsx [hunk 3]
```diff
@@ -154,7 +156,49 @@ export const EventLimitsTab = (props: Pick<EventTypeSetupInfered, "eventType">)
           />
         </div>
       </div>
+
       <hr className="my-8" />
+
+      <div className="flex flex-col space-y-4 lg:flex-row lg:space-y-0 lg:space-x-4">
+        <fieldset className="block flex-col sm:flex">
+          <div className="flex space-x-3">
+            <Controller
```
Comment (sean-brydon) [inline apps/web/components/v2/eventtype/EventLimitsTab.tsx:165]: Toggle active or not -> display PER_DAY option on first toggle

File: apps/web/components/v2/eventtype/EventLimitsTab.tsx [hunk 4]
```diff
@@ -261,3 +305,127 @@ export const EventLimitsTab = (props: Pick<EventTypeSetupInfered, "eventType">)
...
+
+const validationOrderKeys = ["PER_DAY", "PER_WEEK", "PER_MONTH", "PER_YEAR"];
+
+const BookingLimits = () => {
+  const { watch, setValue, control } = useFormContext<FormValues>();
+  const watchBookingLimits = watch("bookingLimits");
+  const animateRef = useRef(null);
+  const { t } = useLocale();
+
+  useEffect(() => {
```
Comment (sean-brydon) [inline apps/web/components/v2/eventtype/EventLimitsTab.tsx:317]: ooo fancy animations

File: apps/web/components/v2/eventtype/EventLimitsTab.tsx [hunk 5]
```diff
@@ -261,3 +305,127 @@ export const EventLimitsTab = (props: Pick<EventTypeSetupInfered, "eventType">)
...
+  return (
+    <Controller
+      name="bookingLimits"
+      control={control}
+      render={({ field: { value, onChange } }) => {
+        return (
+          <ul ref={animateRef}>
+            {value &&
+              watchBookingLimits &&
+              Object.entries(value)
```
Comment (sean-brydon) [inline apps/web/components/v2/eventtype/EventLimitsTab.tsx:352]: Sort in order of day,week,month,year

File: apps/web/components/v2/eventtype/EventLimitsTab.tsx [hunk 6]
```diff
@@ -261,3 +305,127 @@ export const EventLimitsTab = (props: Pick<EventTypeSetupInfered, "eventType">)
...
+                          (option) => !Object.keys(value).includes(option.value)
+                        )}
+                        isSearchable={false}
+                        defaultValue={BOOKING_LIMIT_OPTIONS.find((option) => option.value === key)}
+                        onChange={(val) => {
+                          const current = value;
+                          delete current[bookingLimitKey];
+                          const newData = {
+                            ...current,
+                            [val?.value as keyof BookingLimit]: watchBookingLimits[bookingLimitKey],
```
Comment (sean-brydon) [inline apps/web/components/v2/eventtype/EventLimitsTab.tsx:386]: Change value on select (state management here was trickyyy)

File: apps/web/components/v2/eventtype/EventLimitsTab.tsx [hunk 7]
```diff
@@ -261,3 +305,127 @@ export const EventLimitsTab = (props: Pick<EventTypeSetupInfered, "eventType">)
...
+                })}
+            {value && Object.keys(value).length <= 3 && (
+              <Button
+                color="minimal"
+                StartIcon={Icon.FiPlus}
+                onClick={() => {
+                  if (!value || !watchBookingLimits) return;
+                  const currentKeys = Object.keys(watchBookingLimits);
+
+                  const rest = Object.values(BOOKING_LIMIT_OPTIONS).filter(
```
Comment (sean-brydon) [inline apps/web/components/v2/eventtype/EventLimitsTab.tsx:412]: Figuring out what isn't in the current list

File: apps/web/pages/api/book/event.ts [hunk 8]
```diff
@@ -174,6 +175,7 @@ const getEventTypesFromDB = async (eventTypeId: number) => {
 
   return {
     ...eventType,
+    bookingLimits: parseBookingLimit(eventType.bookingLimits),
```
Comment (sean-brydon) [inline apps/web/pages/api/book/event.ts:178]: Parse booking limits so everything is correctly typed later on

File: apps/web/pages/api/book/event.ts [hunk 9]
```diff
@@ -327,6 +330,11 @@ async function handler(req: NextApiRequest) {
 
   if (!users) throw new HttpError({ statusCode: 404, message: "eventTypeUser.notFound" });
 
+  if (eventType.bookingLimits) {
```
Comment (sean-brydon) [inline apps/web/pages/api/book/event.ts:333]: Validate

File: packages/core/getUserAvailability.ts [hunk 10]
```diff
@@ -129,7 +135,7 @@ export async function getUserAvailability(
     selectedCalendars,
   });
 
-  const bufferedBusyTimes = busyTimes.map((a) => ({
+  const bufferedBusyTimes: EventBusyDetails[] = busyTimes.map((a) => ({
```
Comment (sean-brydon) [inline packages/core/getUserAvailability.ts:138]: Assign this type to `EventBusyDetails[]` as that is what it was -> TS doesn't infer correctly here

File: packages/core/getUserAvailability.ts [hunk 11]
```diff
@@ -139,6 +145,58 @@ export async function getUserAvailability(
...
+    const ourBookings = busyTimes.filter((busyTime) =>
+      busyTime.source?.startsWith(`eventType-${eventType?.id}`)
+    );
+
+    // Apply booking limit filter against our bookings
+
+    for (const [key, limit] of Object.entries(bookingLimits)) {
+      const limitKey = key as keyof BookingLimit;
+
+      if (limitKey === "PER_YEAR") {
```
Comment (sean-brydon) [inline packages/core/getUserAvailability.ts:167]: If PER_YEAR we use DB call to check - as availability only takes in the busy events for that month

File: packages/lib/server/checkBookingLimits.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+import { HttpError } from "../http-error";
+
+export async function checkBookingLimits(
+  bookingLimits: BookingLimit,
+  eventStartDate: Date,
+  eventId: number,
+  returnBusyTimes?: boolean
+) {
+  const limitCalculations = Object.entries(bookingLimits).map(
```
Comment (sean-brydon) [inline packages/lib/server/checkBookingLimits.ts:13]: Loop over all booking limits -> key/value

File: packages/lib/server/checkBookingLimits.ts [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    async ([key, limitingNumber]) =>
+      await checkLimit({ key, limitingNumber, eventStartDate, eventId, returnBusyTimes })
+  );
+  await Promise.all(limitCalculations)
+    .then((res) => {
+      if (returnBusyTimes) {
```
Comment (sean-brydon) [inline packages/lib/server/checkBookingLimits.ts:19]: This is used if we want to return the value - not always the case -> This is used when we calculate availability

File: apps/web/components/v2/eventtype/EventLimitsTab.tsx [hunk 14]
```diff
@@ -261,3 +305,127 @@ export const EventLimitsTab = (props: Pick<EventTypeSetupInfered, "eventType">)
...
+      value: "PER_YEAR",
+      label: "Per Year",
+    },
+  ];
+
+  return (
+    <Controller
+      name="bookingLimits"
+      control={control}
+      render={({ field: { value, onChange } }) => {
```
Comment (JeroenReumkens) [inline apps/web/components/v2/eventtype/EventLimitsTab.tsx:347]: Does it make sense to rename `value` to eg `currentBookingLimits`? Wondering if that makes the code further down better understandable (eg line 382)

File: apps/web/components/v2/eventtype/EventLimitsTab.tsx [hunk 15]
```diff
@@ -261,3 +305,127 @@ export const EventLimitsTab = (props: Pick<EventTypeSetupInfered, "eventType">)
...
+                      />
+                      <Select
+                        options={BOOKING_LIMIT_OPTIONS.filter(
+                          (option) => !Object.keys(value).includes(option.value)
+                        )}
+                        isSearchable={false}
+                        defaultValue={BOOKING_LIMIT_OPTIONS.find((option) => option.value === key)}
+                        onChange={(val) => {
+                          const current = value;
+                          delete current[bookingLimitKey];
```
Comment (JeroenReumkens) [inline apps/web/components/v2/eventtype/EventLimitsTab.tsx:383]:
```suggestion
                          // Removes limit from previous selected value (eg when changed from per_week to per_month, we unset per_week here)
                          delete current[bookingLimitKey];
```

File: apps/web/components/v2/eventtype/EventLimitsTab.tsx (hunk 6 above)
Comment (JeroenReumkens) [inline apps/web/components/v2/eventtype/EventLimitsTab.tsx:386]:
```suggestion
                            // Set limit to new selected value (in the example above this means we set the limit to per_week here).
                            [val?.value as keyof BookingLimit]: watchBookingLimits[bookingLimitKey],
```

File: packages/lib/server/checkBookingLimits.ts [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+        return {
+          start: startDate,
+          end: endDate,
+        };
+      }
+
+      throw new HttpError({
+        message: `Booking limit of ${limitingNumber}/${filter} reached for this eventType`,
+        statusCode: 401,
```
Comment (JeroenReumkens) [inline packages/lib/server/checkBookingLimits.ts:85]: Is 401 the right statuscode here?

Comment (sean-brydon) [reply packages/lib/server/checkBookingLimits.ts:85]: Good shout - should be 403 (forbidden) for if we wanna go super fancy use 422 (unprocurable entity )

Comment (sean-brydon) [reply packages/lib/server/checkBookingLimits.ts:85]: Setting to 403 for consistency

**Review by PeerRich — COMMENTED** [review]
not a huge fan of throwing so many toasts. I wonder if there is a better UI solution rather than showing errors etc.

it's a bit of a guessing game right now @Jaibles 

@JeroenReumkens maybe you have an idea, too

**Comment by sean-brydon** [comment]
>  [  ](https://user-images.githubusercontent.com/55134778/193813865-23d9185d-2e05-4d69-a1ea-1e0e3775c613.gif) [ ](https://user-images.githubusercontent.com/55134778/193813865-23d9185d-2e05-4d69-a1ea-1e0e3775c613.gif)
> 
> not a huge fan of throwing so many toasts. I wonder if there is a better UI solution rather than showing errors etc.
> 
> it's a bit of a guessing game right now @Jaibles
> 
> @JeroenReumkens maybe you have an idea, too

Note - this is only on event type saving (pressing enter or clicking save) and not as the user is changing the field values

**Review by pumfleet — APPROVED** [review]
Code looks good to me

File: apps/web/pages/v2/event-types/[type]/index.tsx [hunk 17]
```diff
@@ -11,12 +11,13 @@ import { z } from "zod";
...
 import getApps, { getLocationOptions } from "@calcom/app-store/utils";
 import { LocationObject, EventLocationType } from "@calcom/core/location";
-import { parseRecurringEvent } from "@calcom/lib";
+import { parseRecurringEvent, parseBookingLimit, validateBookingLimitOrder } from "@calcom/lib";
 import { CAL_URL } from "@calcom/lib/constants";
 import { useLocale } from "@calcom/lib/hooks/useLocale";
 import prisma from "@calcom/prisma";
-import { trpc } from "@calcom/trpc/react";
-import type { RecurringEvent } from "@calcom/types/Calendar";
+import { trpc, TRPCClientError } from "@calcom/trpc/react";
```
Comment (zomars) [inline apps/web/pages/v2/event-types/[type]/index.tsx:18]:
```suggestion
import { trpc } from "@calcom/trpc/react";
```

File: apps/web/pages/v2/event-types/[type]/index.tsx [hunk 18]
```diff
@@ -11,12 +11,13 @@ import { z } from "zod";
...
+import { TRPCError } from "@calcom/trpc/server";
```
Comment (zomars) [inline apps/web/pages/v2/event-types/[type]/index.tsx:19]:
```suggestion
```

File: apps/web/test/lib/checkBookingLimits.test.ts [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+
+const MOCK_DATA: Mockdata = {
+  id: 1,
+  startDate: dayjs("2022-09-30T09:00:00+01:00").toDate(),
+  bookingLimits: {
+    PER_DAY: 1,
+  },
+};
+
+describe("Check Booking Limits Tests", () => {
```
Comment (zomars) [inline apps/web/test/lib/checkBookingLimits.test.ts:22]: https://www.youtube.com/watch?v=lg5WKsVnEA4

**Comment by hariombalhara** [comment]
Amazing work @sean-brydon 🙌. That's a lot of validations going on.

Do we have improvements in troubleshooter planned for this or maybe already done? Because 1 booking can block the entire day, week or month, it should be shown as such in Troubleshooter

**Comment by sean-brydon** [comment]
> Amazing work @sean-brydon 🙌. That's a lot of validations going on.
> 
> Do we have improvements in troubleshooter planned for this or maybe already done? Because 1 booking can block the entire day, week or month, it should be shown as such in Troubleshooter

Thank you! I will create an issue for this as it's a very good point! Never crossed my mind

File: packages/types/Calendar.d.ts [hunk 20]
```diff
@@ -105,6 +105,13 @@ export interface RecurringEvent {
   tzid?: string | undefined;
 }
 
+export interface BookingLimit {
```
Comment (hariombalhara) [inline packages/types/Calendar.d.ts:108]: nit: To make code more DRY
```suggestion
export type BookingLimit = z.infer<typeof bookingLimitsType>;
```

File: packages/lib/server/checkBookingLimits.ts [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+import { HttpError } from "../http-error";
+
+export async function checkBookingLimits(
+  bookingLimits: BookingLimit,
+  eventStartDate: Date,
+  eventId: number,
+  returnBusyTimes?: boolean
+) {
+  const limitCalculations = Object.entries(bookingLimits).map(
+    async ([key, limitingNumber]) =>
```
Comment (hariombalhara) [inline packages/lib/server/checkBookingLimits.ts:14]: I think this async, await is unnecessary as limitCalculations is expected to be an array of promises which can be achieved as 

```js
Object.entries(bookingLimits).map(
    ([key, limitingNumber]) =>
      checkLimit({ key, limitingNumber, eventStartDate, eventId, returnBusyTimes })
  );
```

File: packages/trpc/server/routers/viewer/eventTypes.tsx [hunk 22]
```diff
@@ -296,6 +298,13 @@ export const eventTypesRouter = createProtectedRouter()
         data.customInputs = handleCustomInputs(customInputs, id);
       }
 
+      if (bookingLimits) {
+        const isValid = validateBookingLimitOrder(bookingLimits);
+        if (!isValid)
+          throw new TRPCError({ code: "BAD_REQUEST", message: "Booking limits must be in ascending order." });
```
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/eventTypes.tsx:304]: nit: 
`validateBookingLimitOrder` theoretically can check for any other issues, so it knows best what error should be thrown.

**Review by hariombalhara — APPROVED** [review]

**Comment by PeerRich** [comment]
@sean-brydon lets merge this but follow up with the changes by @hariombalhara

**Review by zomars — DISMISSED** [review]
Blocking since there are some type check errors

image

**Comment by zomars** [comment]
> @zomars, @sean-brydon is OOO for 7 days. should someone else fix those?

@alannnc said he would take this one

**Review by alannnc — APPROVED** [review]
Now type checks pass and functionality works as expected. Should be ready to test it on cal.dev before going to production.

**Comment by hariombalhara** [comment]
Resolved Deployment issue. Should be merged automatically now.

**Pull request merged**